lighthouse-core/gather/gatherers/seo/tap-targets.js

```diff
@@ -0,0 +1,351 @@
+    return false;
+  }
+
+  return true;
+}
+
+/**
+ * @param {LH.Artifacts.Rect[]} clientRects
+ */
+/* istanbul ignore next */
+function allClientRectsEmpty(clientRects) {
+  return clientRects.every(cr => cr.width === 0 && cr.height === 0);
+}
+
+/**
+ * @param {Element} node
+ */
+/* istanbul ignore next */
+function getVisibleClientRects(node) {
+  if (!nodeIsVisible(node)) {
+    return [];
+  }
+
+  const {
+    overflowX,
+    overflowY,
+  } = getComputedStyle(node);
+  let clientRects = getClientRects(node, true);
+
+  if (allClientRectsEmpty(clientRects)) {
+    if ((overflowX === 'hidden' && overflowY === 'hidden') || node.children.length === 0) {
+      // own size is 0x0 and there's no visible child content
+      return [];
+    }
+  }
+
+  // Treating overflowing content in scroll containers as invisible could mean that
+  // most of a given page is deemed invisible. But:
+  // - tap targets audit doesn't consider different containers/layers
+  // - having most content in an explicit scroll container is rare
+  // - treating them as hidden only generates false passes, which is better than false failures
+  clientRects = filterClientRectsWithinAncestorsVisibleScrollArea(node, clientRects);
+
+  return clientRects;
+}
+
+/**
+ *
+ * @param {Element} node
+ * @param {LH.Artifacts.Rect[]} clientRects
+ * @returns {LH.Artifacts.Rect[]}
+ */
+/* istanbul ignore next */
+function filterClientRectsWithinAncestorsVisibleScrollArea(node, clientRects) {
+  const parent = node.parentElement;
+  if (!parent) {
+    return clientRects;
+  }
+  if (getComputedStyle(parent).overflowY !== 'visible') {
+    const parentBCR = parent.getBoundingClientRect();
+    clientRects = clientRects.filter(cr => rectContains(parentBCR, cr));
+  }
+  if (parent.parentElement && parent.parentElement.tagName !== 'BODY') {
+    return filterClientRectsWithinAncestorsVisibleScrollArea(
+      parent,
+      clientRects
+    );
+  }
+  return clientRects;
+}
+
+/**
+ * @param {Element} node
+ * @param {boolean} includeChildren
+ * @returns {LH.Artifacts.Rect[]}
+ */
+/* istanbul ignore next */
+function getClientRects(node, includeChildren = true) {
+  /** @type {LH.Artifacts.Rect[]} */
+  let clientRects = Array.from(
+    node.getClientRects()
+  ).map(clientRect => {
+    // Contents of DOMRect get lost when returned from Runtime.evaluate call,
+    // so we convert them to plain objects.
+    const {width, height, left, top, right, bottom} = clientRect;
+    return {width, height, left, top, right, bottom};
+  });
+  if (includeChildren) {
+    for (const child of node.children) {
+      clientRects = clientRects.concat(getClientRects(child));
+    }
+  }
+
+  return clientRects;
+}
+
+/**
+ * @param {Element} node
+ * @returns {boolean}
+ */
+/* istanbul ignore next */
+function nodeHasParentTapTarget(node) {
+  if (!node.parentElement) {
+    return false;
+  }
+  if (node.parentElement.matches(tapTargetsSelector)) {
```

Comment:
Cool, changed that! Should we add some kind of test case for when `matches` is overwritten?

Reply:
>  Should we add some kind of test case for when matches is overwritten?

We do have [the dbw smoke test check that it's possible to work around the issue](https://github.com/GoogleChrome/lighthouse/blob/8ed55613d41e47e44d2be032e671571177aa5c53/lighthouse-cli/test/fixtures/dobetterweb/dbw_tester.html#L124), but we'd have to add it to the SEO smoke test as well to catch regressions in this use. However...

> actually is it possible for us to just `useIsolation` here? 

this is a good idea and lets the code skip using the cached version. [Example in `link-elements.js`](https://github.com/GoogleChrome/lighthouse/blob/8ed55613d41e47e44d2be032e671571177aa5c53/lighthouse-core/gather/gatherers/link-elements.js#L32).

Reply:
Nice, `useIsolation` works. And once we update the sample json it will run the tap targets gatherer there too and would pick it up if we broke this in the future?

Reply:
And currently there's no full E2E LH run where we could overwrite `matches`? If that's the case I'll add it to the `seo-tap-targets.html`.

Reply:
The smokehouse tests are the only E2E tests we have, so that's the right place to put such a test 👍